pkg/etcdcli/health.go

```diff
@@ -0,0 +1,154 @@
+			continue
+		}
+		wg.Add(1)
+		go func(member *etcdserverpb.Member) {
+			defer wg.Done()
+			// new client vs shared is used here to minimize disruption of cached client consumers.
+			// performance analisis of CPU and RSS consumption showed net gain after refactor
+			cli, err := getEtcdClient([]string{member.ClientURLs[0]})
+			if err != nil {
+				hch <- healthCheck{Member: member, Healthy: false, Started: true, Error: fmt.Errorf("create client failure: %w", err)}
+				return
+			}
+			defer cli.Close()
+			st := time.Now()
+			ctx, cancel := context.WithCancel(context.Background())
+			// linearized request to verify health of member
+			_, err = cli.Get(ctx, "health")
+			cancel()
+			hc := healthCheck{Member: member, Healthy: false, Started: true, Took: time.Since(st).String()}
+			if err == nil || err == rpctypes.ErrPermissionDenied {
+				hc.Healthy = true
+			} else {
+				hc.Error = fmt.Errorf("health check failed: %w", err)
+			}
+			hch <- hc
+		}(member)
+	}
+
+	wg.Wait()
+	close(hch)
+
+	for healthCheck := range hch {
+		memberHealth = append(memberHealth, healthCheck)
+	}
+
+	return memberHealth
+}
+
+// Status returns a reporting of memberHealth status
+func (h memberHealth) Status() string {
+	healthyMembers := h.GetHealthyMemberNames()
+
+	status := []string{}
+	if len(h) == len(healthyMembers) {
+		status = append(status, fmt.Sprintf("%d members are available", len(h)))
+	} else {
+		status = append(status, fmt.Sprintf("%d of %d members are available", len(healthyMembers), len(h)))
+		for _, etcd := range h {
+			switch {
+			case !etcd.Started:
+				status = append(status, fmt.Sprintf("%s has not started", GetMemberNameOrHost(etcd.Member)))
+				continue
+			case !etcd.Healthy:
+				status = append(status, fmt.Sprintf("%s is unhealthy", etcd.Member.Name))
+				continue
+			}
+		}
+	}
+	return strings.Join(status, ", ")
+}
+
+// Healthy returns healthy members
+func (h memberHealth) GetHealthyMembers() []*etcdserverpb.Member {
+	members := []*etcdserverpb.Member{}
+	for _, etcd := range h {
+		if etcd.Healthy {
+			members = append(members, etcd.Member)
+		}
+	}
+	return members
+}
+
+// HealthyMemberNames returns a list of healthy member names
+func (h memberHealth) GetHealthyMemberNames() []string {
+	memberNames := []string{}
+	for _, etcdMember := range h.GetHealthyMembers() {
+		memberNames = append(memberNames, etcdMember.Name)
+	}
+	return memberNames
+}
+
+// Unhealthy returns unhealthy members
+func (h memberHealth) GetUnhealthyMembers() []*etcdserverpb.Member {
+	members := []*etcdserverpb.Member{}
+	for _, etcd := range h {
+		if !etcd.Healthy {
+			members = append(members, etcd.Member)
+		}
+	}
+	return members
+}
+
+// UnhealthyMemberNames returns a list of unhealthy member names
+func (h memberHealth) GetUnhealthyMemberNames() []string {
+	memberNames := []string{}
+	for _, etcdMember := range h.GetUnhealthyMembers() {
+		if len(etcdMember.ClientURLs) == 0 {
```

Comment:
nit: this is kind of weird, seems like it should be part of GetMemberNameOrHost or something. Seems smelly that callers have to make some decisions to compute the name of a Member. Would be nice to encapsulate the name logic in a single place (so callers always access a member name through said function) or something.